// @debridge-finance/dln-client/dist/esm/evm/core/client/dln-evm.client.js
import Web3 from "web3";
import { helpers } from "@debridge-finance/solana-utils";
import { SupportedEVMChains } from "../types/common.type";
import { DlnSourceService } from "../services/dln-source.service";
import { DlnDestinationService } from "../services/dln-destination.service";
import { ClientError } from "../../../errors/client-error.error";
import { ClientErrorType } from "../../../errors/client-error-type.enum";
import { DebridgeGateService } from "../services/debridge-gate.service";
import { BPS_DENOMINATOR, ChainEngine, ChainId, getEngineByChainId, } from "../../../common.types";
import { ServiceType } from "../enums/service-type.enum";
import { getEvmAccountBalance, getEvmDecimals, getEvmTokenName, getEvmTokenSymbol } from "../utils/get.evm.info";
import { CrossChainForwarderService } from "../services/crosschain-forwarder.service";
import { Logger } from "../../../common/logger";
import { orderToEvm } from "../utils/order.data.to.evm";
import { buffersAreEqual, calcBigIntAmountViaCrossRate } from "../../../utils";
import { getOrderFromEvmTransaction } from "../utils/get.order.from.transaction";
import { OrderGiveStatus } from "../models/give-orders.model";
import { Env } from "../../../common/env";
export class DlnClient {
    getEngine() {
        return ChainEngine.EVM;
    }
    packCalldata(params, payload) {
        const envelopeVersion = 1;
        const web3 = new Web3();
        const encodedTuple = web3.eth.abi.encodeParameters([
            {
                type: "tuple",
                name: "ExternalCallEnvelopV1",
                components: [
                    { name: "fallbackAddress", type: "address" },
                    { name: "executorAddress", type: "address" },
                    { name: "executionFee", type: "uint160" },
                    { name: "allowDelayedExecution", type: "bool" },
                    { name: "requireSuccessfullExecution", type: "bool" },
                    { name: "payload", type: "bytes" },
                ],
            },
        ], [
            {
                fallbackAddress: this.fromAddress(params.fallbackAddress),
                executorAddress: this.fromAddress(payload.executorAddress),
                executionFee: params.executionFee.toString(),
                allowDelayedExecution: params.allowDelayedExecution,
                requireSuccessfullExecution: params.requireSuccessfullExecution,
                payload: helpers.bufferToHex(params.payload),
            },
        ]);
        // chainEngine uint8, envelope version uint8, bytes envelope
        const data = web3.utils.encodePacked({ value: envelopeVersion.toString(), type: "uint8" }, { value: encodedTuple, type: "bytes" });
        if (data === null)
            throw new Error("Failed to encode!");
        return data;
    }
    unpackCalldata(calldata) {
        const web3 = new Web3();
        const version = calldata[0];
        const decoded = web3.eth.abi.decodeParameters([
            {
                type: "tuple",
                name: "ExternalCallEnvelopV1",
                components: [
                    { name: "fallbackAddress", type: "address" },
                    { name: "executorAddress", type: "address" },
                    { name: "executionFee", type: "uint160" },
                    { name: "allowDelayedExecution", type: "bool" },
                    { name: "requireSuccessfullExecution", type: "bool" },
                    { name: "payload", type: "bytes" },
                ],
            },
        ], calldata.subarray(1).toString("hex"))[0];
        return {
            version,
            data: {
                allowDelayedExecution: decoded.allowDelayedExecution,
                requireSuccessfullExecution: decoded.requireSuccessfullExecution,
                payload: helpers.hexToBuffer(decoded.payload),
                executionFee: BigInt(decoded.executionFee),
                fallbackAddress: this.parseAddress(decoded.fallbackAddress),
                executorAddress: this.parseAddress(decoded.executorAddress),
            },
        };
    }
    constructor(config, version = Env.Prod) {
        this.config = config;
        this.version = version;
        /**
         * @internal
         */
        this.SOLANA_DENOMINATOR_DELTA = 8;
        this.nativeCurrencyDecimals = 18;
        this.addressLength = 20;
        this.cache = new Map();
        this.connectionManager = new Map();
        for (const [chain, chainConfig] of Object.entries(config.chainConfig)) {
            if (chainConfig.connection !== undefined)
                this.connectionManager.set(parseInt(chain), chainConfig.connection);
        }
        this.dlnSourceService = new DlnSourceService(this.config.enableContractsCache, version);
        this.dlnDestinationService = new DlnDestinationService(this.config.enableContractsCache, version);
        this.debridgeGateService = new DebridgeGateService(this.config.enableContractsCache);
        this.crosschainForwarderService = new CrossChainForwarderService();
        this.defaultLogger = new Logger();
    }
    getConnection(chain) {
        const web3 = this.connectionManager.get(chain);
        if (!web3)
            throw new ClientError(ClientErrorType.ConnectionMissing, `Failed to get connection for chain ${chain}`);
        return web3;
    }
    setConnection(chain, connection) {
        this.connectionManager.set(chain, connection);
    }
    get nativeAddress() {
        return Uint8Array.from(Array.from({ length: this.addressLength }).fill(0));
    }
    fromAddress(address) {
        return helpers.bufferToHex(address);
    }
    toAddress(nativeAddress) {
        return helpers.hexToBuffer(nativeAddress);
    }
    parseAddress(address) {
        return this.toAddress(address);
    }
    getForwarderAddress(chain) {
        return this.toAddress(this.getContractAddress(chain, ServiceType.CrosschainForwarder));
    }
    async getWrappedNativeTokenAddress(chain) {
        const cacheKey = `getWrappedNativeTokenAddress_${chain}`;
        if (this.cache.has(cacheKey)) {
            return this.toAddress(this.cache.get(cacheKey));
        }
        const address = this.getContractAddress(chain, ServiceType.DebridgeGate);
        const weth = await this.debridgeGateService.weth(this.getConnection(chain), address);
        this.cache.set(cacheKey, weth);
        return this.toAddress(weth);
    }
    buildCacheKey(prefix, chain, address) {
        return `${prefix}_${chain}_${this.fromAddress(address)}`;
    }
    async getDecimals(chain, tokenAddress) {
        const cacheKey = this.buildCacheKey("decimals", chain, tokenAddress);
        const cacheHit = this.cache.get(cacheKey);
        if (cacheHit) {
            return cacheHit;
        }
        else {
            const decimals = await getEvmDecimals(this.fromAddress(tokenAddress), this.getConnection(chain));
            this.cache.set(cacheKey, decimals);
            return decimals;
        }
    }
    async getTokenInfo(chain, tokenAddress) {
        const cacheKey = this.buildCacheKey("tokenInfo", chain, tokenAddress);
        const cacheHit = this.cache.get(cacheKey);
        if (cacheHit) {
            return cacheHit;
        }
        else {
            const provider = this.getConnection(chain);
            const stringifiedAddress = this.fromAddress(tokenAddress);
            const [decimals, name, symbol] = await Promise.all([
                this.getDecimals(chain, tokenAddress),
                getEvmTokenName(stringifiedAddress, chain, provider),
                getEvmTokenSymbol(stringifiedAddress, chain, provider),
            ]);
            const result = {
                address: stringifiedAddress,
                chainId: chain,
                decimals,
                name,
                symbol,
            };
            this.cache.set(cacheKey, result);
            return result;
        }
    }
    patchOrderGive(params, payload) {
        const address = this.getContractAddress(params.order.give.chainId, ServiceType.Source);
        const evmOrder = orderToEvm(params.order);
        const method = this.dlnSourceService.getPatchOrderGiveMethod(this.getConnection(params.order.give.chainId), address, evmOrder, params.amountToAdd.toString(), payload?.permit || "0x");
        let value = BigInt(0);
        if (buffersAreEqual(params.order.give.tokenAddress, this.nativeAddress)) {
            value = params.amountToAdd;
        }
        const tx = {
            data: method.encodeABI(),
            to: address,
            value: value.toString(),
        };
        return Promise.resolve(tx);
    }
    patchOrderTake(params, payload) {
        const address = this.getContractAddress(params.order.take.chainId, ServiceType.Destination);
        const method = this.dlnDestinationService.getPatchOrderTakeMethod(this.getConnection(params.order.take.chainId), address, orderToEvm(params.order), params.amountToSub.toString());
        const tx = {
            data: method.encodeABI(),
            to: address,
            value: 0,
        };
        return Promise.resolve(tx);
    }
    async getFinalGiveAmount(params, payload) {
        const address = this.getContractAddress(params.order.give.chainId, ServiceType.Source);
        const patches = await this.dlnSourceService.givePatches(this.getConnection(params.order.give.chainId), address, params.order.orderId);
        return params.order.give.amount + patches;
    }
    async getFinalTakeAmount(params, payload) {
        const address = this.getContractAddress(params.order.take.chainId, ServiceType.Destination);
        const patches = await this.dlnDestinationService.takePatches(this.getConnection(params.order.take.chainId), address, params.order.orderId);
        return params.order.take.amount - patches;
    }
    /**
     * @internal
     */
    async preswapAndCreateOrder(params, payload) {
        const currentChain = params.order.give.chainId;
        const { order, swapResult } = params;
        if ("evm" !== swapResult.engine) {
            throw new ClientError(ClientErrorType.PropertyMissing, "Wrong swap response provided, missing 'solana' engine in swapResponse");
        }
        const provider = this.getConnection(currentChain);
        const { tx: { data: srcSwapCalldata, to: srcSwapRouter }, fromToken, toToken: pickedGiveToken, amountIn: giveTokenAmount, } = swapResult;
        const receiverCalldata = (await this.createOrder({ ...params, order }, { permit: "0x" })).data;
        const [sourceGlobalFixedNativeFee] = await Promise.all([
            this.getSourceGlobalFixedNativeFee(currentChain, provider),
        ]);
        let value = BigInt(sourceGlobalFixedNativeFee);
        if (buffersAreEqual(fromToken, this.nativeAddress)) {
            value += giveTokenAmount;
        }
        const method = this.crosschainForwarderService.strictlySwapAndCall(provider, this.fromAddress(payload.sender), {
            _srcTokenIn: this.fromAddress(fromToken),
            _srcAmountIn: giveTokenAmount,
            _srcTokenInPermitEnvelope: payload.srcTokenInPermitEnvelope,
            _srcSwapRouter: srcSwapRouter,
            _srcSwapCalldata: srcSwapCalldata,
            _srcTokenOut: this.fromAddress(pickedGiveToken),
            _srcTokenExpectedAmountOut: order.give.amount,
            _srcTokenRefundRecipient: this.fromAddress(payload.srcTokenRefundRecipient),
            _receiver: this.getContractAddress(currentChain, ServiceType.Source),
            _receiverCalldata: receiverCalldata,
        });
        return {
            value: value.toString(),
            data: method,
            to: this.getContractAddress(currentChain, ServiceType.CrosschainForwarder),
        };
    }
    /**
     * @internal
     */
    async preswapAndFulfillOrder(params, payload) {
        const currentChain = params.order.take.chainId;
        const { data: receiverCalldata } = await this.fulfillOrder(params, {
            unlockAuthority: payload.unlockAuthority,
            permit: "0x",
            externalCallRewardBeneficiary: payload.externalCallRewardBeneficiary,
        });
        const { order } = params;
        if ("evm" !== params.swapResult.engine) {
            throw new ClientError(ClientErrorType.PropertyMissing, "Wrong swap response provided, missing 'solana' engine in swapResponse");
        }
        const { tx: { data: swapCalldata, to: srcSwapRouter }, fromToken, amountIn, } = params.swapResult;
        let value = BigInt(0);
        if (buffersAreEqual(fromToken, this.nativeAddress)) {
            value += amountIn;
        }
        const method = this.crosschainForwarderService.strictlySwapAndCallDln({
            _srcTokenIn: this.fromAddress(fromToken),
            _srcAmountIn: amountIn,
            _srcTokenInPermitEnvelope: payload.srcTokenInPermitEnvelope || "0x",
            _srcSwapRouter: srcSwapRouter,
            _srcSwapCalldata: swapCalldata,
            _srcTokenOut: this.fromAddress(order.take.tokenAddress),
            _srcTokenExpectedAmountOut: order.take.amount,
            // refund from fulfill's preswap is sent to the user (the maker) to limit taker's income
            _srcTokenRefundRecipient: this.fromAddress(payload.preswapChangeRecipient || order.receiver),
            _receiver: this.getContractAddress(currentChain, ServiceType.Destination),
            _receiverCalldata: receiverCalldata,
            _orderId: helpers.bufferToHex(order.orderId),
        });
        return {
            value: value.toString(),
            data: method,
            to: this.getContractAddress(currentChain, ServiceType.CrosschainForwarder),
        };
    }
    getBalance(chain, tokenAddress, wallet) {
        return getEvmAccountBalance(this.fromAddress(tokenAddress), this.fromAddress(wallet), this.getConnection(chain));
    }
    async getFees(chain, amount) {
        const web3 = this.getConnection(chain);
        const fixFee = await this.getSourceGlobalFixedNativeFee(chain, web3);
        const transferFee = await this.getSourceGlobalTransferFeeBps(chain, web3);
        return {
            fixFee: BigInt(fixFee),
            transferFee: amount ? (amount * BigInt(transferFee)) / BPS_DENOMINATOR : BigInt(0),
            transferFeeBps: Number(transferFee),
        };
    }
    async getSourceGlobalFixedNativeFee(chainId, web3) {
        const key = `getSourceGlobalFixedNativeFee_${chainId}`;
        if (this.cache.has(key)) {
            return this.cache.get(key);
        }
        const address = this.getContractAddress(chainId, ServiceType.Source);
        const result = await this.dlnSourceService.getGlobalFixedNativeFee(web3, address);
        this.cache.set(key, result);
        return result;
    }
    async getSourceGlobalTransferFeeBps(chainId, web3) {
        const key = `getSourceGlobalTransferFeeBps_${chainId}`;
        if (this.cache.has(key)) {
            return this.cache.get(key);
        }
        const address = this.getContractAddress(chainId, ServiceType.Source);
        const result = this.dlnSourceService.getGlobalTransferFeeBps(web3, address);
        this.cache.set(key, result);
        return result;
    }
    async getGlobalTransferFeeBps(chainId, web3) {
        const key = `getGlobalTransferFeeBps_${chainId}`;
        if (this.cache.has(key)) {
            return Number(this.cache.get(key));
        }
        const data = await this.debridgeGateService.getGlobalTransferFeeBps(web3, this.getContractAddress(chainId, ServiceType.DebridgeGate));
        this.cache.set(key, data);
        return Number(data);
    }
    async getGlobalFixedNativeFee(chainId, web3) {
        const key = `getGlobalFixedNativeFee_${chainId}`;
        if (this.cache.has(key)) {
            return BigInt(this.cache.get(key));
        }
        const res = await this.debridgeGateService.getGlobalFixedNativeFee(web3, this.getContractAddress(chainId, ServiceType.DebridgeGate));
        this.cache.set(key, res);
        return BigInt(res);
    }
    async getAmountWithAccountedGateTransferFee(sendFromChainId, neededAmount, web3) {
        const transferFee = await this.getGlobalTransferFeeBps(sendFromChainId, web3);
        return (neededAmount * BPS_DENOMINATOR) / (BPS_DENOMINATOR - BigInt(transferFee));
    }
    async getValueWithIncludedFee(executionFee, chainId, web3) {
        executionFee = BigInt(executionFee.toString());
        const globalFixedNativeFee = await this.getGlobalFixedNativeFee(chainId, web3);
        const accountedExecutionFee = await this.getAmountWithAccountedGateTransferFee(chainId, executionFee, web3);
        const resultingValue = globalFixedNativeFee + accountedExecutionFee;
        return resultingValue;
    }
    async getValueWithIncludedFeeAndRewards(rewards, executionFee, chainId, web3) {
        executionFee = BigInt(executionFee.toString());
        const rewardSum = rewards.reduce((acc, reward) => acc + BigInt(reward.toString()), BigInt(0)) *
            BigInt(10) ** BigInt(this.nativeCurrencyDecimals - this.SOLANA_DENOMINATOR_DELTA);
        const rewardAndExecutionFee = executionFee + rewardSum;
        return this.getValueWithIncludedFee(rewardAndExecutionFee, chainId, web3);
    }
    async getCreateCost(params, payload) {
        return Promise.reject("Not implemented");
    }
    async getGiveOrderState(params, payload) {
        const contract = this.getContractAddress(params.giveChain, ServiceType.Source);
        const web3 = this.getConnection(params.giveChain);
        const result = await this.dlnSourceService.giveOrder(web3, contract, params.orderId);
        if (result.status === OrderGiveStatus.NotSet)
            return null;
        const statusMap = {
            [OrderGiveStatus.Created]: 1 /* OrderState.Created */,
            [OrderGiveStatus.ClaimedCancel]: 7 /* OrderState.ClaimedCancel */,
            [OrderGiveStatus.ClaimedUnlock]: 6 /* OrderState.ClaimedUnlock */,
        };
        if (!Object.keys(statusMap).includes(result.status.toString()))
            throw new ClientError(ClientErrorType.UnknownStatus, `unknown status ${result.status}`);
        return {
            status: statusMap[result.status],
        };
    }
    async getTakeOrderState(params, payload) {
        const contract = this.getContractAddress(params.takeChain, ServiceType.Destination);
        const web3 = this.getConnection(params.takeChain);
        const result = await this.dlnDestinationService.takeOrders(web3, contract, params.orderId);
        if (result.status === "0")
            return null;
        const statusMap = {
            1: 2 /* OrderState.Fulfilled */,
            2: 3 /* OrderState.SentUnlock */,
            3: 5 /* OrderState.SentCancel */,
        };
        if (!Object.keys(statusMap).includes(result.status.toString()))
            throw new ClientError(ClientErrorType.UnknownStatus, `unknown take status ${result.status}`); // unknown status
        return {
            giveChainId: Number(result.giveChainId),
            status: statusMap[result.status],
            takerAddress: this.toAddress(result.takerAddress),
        };
    }
    async createOrder(params, payload) {
        const { order, referralCode, affiliateFee } = params;
        const web3 = this.getConnection(order.give.chainId);
        let affiliateFeeStruct = "0x";
        if (affiliateFee !== undefined) {
            affiliateFeeStruct = web3.utils.encodePacked({ value: this.fromAddress(affiliateFee.beneficiary), type: "address" }, { value: affiliateFee.amount.toString(), type: "uint256" });
        }
        const address = this.getContractAddress(order.give.chainId, ServiceType.Source);
        let value = BigInt(await this.getSourceGlobalFixedNativeFee(order.give.chainId, web3));
        if (buffersAreEqual(order.give.tokenAddress, this.nativeAddress)) {
            value += order.give.amount;
        }
        let method;
        const evmOrder = orderToEvm({
            ...order,
            nonce: undefined,
        });
        if (Object.prototype.hasOwnProperty.call(order, "nonce") && order.nonce !== undefined) {
            method = this.dlnSourceService.getCreateSaltedOrderMethod(web3, address, evmOrder, params.order.nonce, affiliateFeeStruct, referralCode || 0, payload.permit, params.metadata || "0x");
        }
        else {
            method = this.dlnSourceService.getCreateOrderMethod(web3, address, evmOrder, affiliateFeeStruct, referralCode || 0, payload.permit);
        }
        return {
            data: method.encodeABI(),
            to: address,
            value: value.toString(),
        };
    }
    async fulfillOrder(params, payload) {
        const { order } = params;
        const web3 = this.getConnection(order.take.chainId);
        const address = this.getContractAddress(order.take.chainId, ServiceType.Destination);
        const method = this.dlnDestinationService.getFulfillOrderMethod(web3, address, order, (payload.fulFillAmount || (await this.getFinalTakeAmount(params, {}))).toString(), order.orderId, payload.permit, this.fromAddress(payload.unlockAuthority), payload.externalCallRewardBeneficiary);
        let value = BigInt(0);
        if (buffersAreEqual(order.take.tokenAddress, this.nativeAddress)) {
            value += order.take.amount;
        }
        return Promise.resolve({
            data: method.encodeABI(),
            to: address,
            value: value.toString(),
        });
    }
    async sendCancelOrder(params, payload) {
        const { order, executionFee, cancelBeneficiary } = params;
        const web3 = this.getConnection(order.take.chainId);
        const address = this.getContractAddress(order.take.chainId, ServiceType.Destination);
        switch (getEngineByChainId(order.give.chainId)) {
            case ChainEngine.EVM: {
                const method = this.dlnDestinationService.getSendEvmOrderCancelMethod(web3, address, orderToEvm(order), this.fromAddress(cancelBeneficiary), executionFee.toString());
                const value = await this.getValueWithIncludedFee(executionFee, order.take.chainId, web3);
                return {
                    data: method.encodeABI(),
                    to: address,
                    value: value.toString(),
                };
            }
            case ChainEngine.Solana: {
                if (cancelBeneficiary.length != 32)
                    throw new ClientError(ClientErrorType.WrongSolanaAddressLength);
                if (payload?.rewards === undefined)
                    throw new Error("No rewards provided");
                const method = this.dlnDestinationService.getSendSolanaOrderCancelMethod(web3, address, orderToEvm(order), this.fromAddress(cancelBeneficiary), executionFee.toString(), payload.rewards[0].toString(), payload.rewards[1].toString());
                const value = await this.getValueWithIncludedFeeAndRewards(payload.rewards, executionFee, order.take.chainId, web3);
                const result = {
                    data: method.encodeABI(),
                    to: address,
                    value: value.toString(),
                };
                return result;
            }
        }
    }
    async sendUnlockOrder(params, payload) {
        const { order, beneficiary, executionFee } = params;
        const address = this.getContractAddress(order.take.chainId, ServiceType.Destination);
        const web3 = this.getConnection(order.take.chainId);
        const encodedBeneficiary = helpers.bufferToHex(beneficiary);
        switch (getEngineByChainId(order.give.chainId)) {
            case ChainEngine.EVM: {
                const method = this.dlnDestinationService.getSendEvmUnlockMethod(web3, address, order.orderId, encodedBeneficiary, executionFee.toString());
                const value = await this.getValueWithIncludedFee(executionFee.toString(), order.take.chainId, web3);
                return {
                    data: method.encodeABI(),
                    value: value.toString(),
                    to: address,
                };
            }
            case ChainEngine.Solana: {
                if (beneficiary.length != 32)
                    throw new ClientError(ClientErrorType.WrongSolanaAddressLength);
                const method = this.dlnDestinationService.getSendSolanaUnlockMethod(web3, address, orderToEvm(order), encodedBeneficiary, executionFee.toString(), payload.rewards[0].toString(), payload.rewards[1].toString());
                const value = await this.getValueWithIncludedFeeAndRewards(payload.rewards, executionFee, order.take.chainId, web3);
                return {
                    data: method.encodeABI(),
                    value: value.toString(),
                    to: address,
                };
            }
        }
    }
    async sendBatchUnlock(params, payload) {
        const { orders, beneficiary, executionFee } = params;
        const takeChain = orders[0].take.chainId;
        const { tokenAddress: giveToken, chainId: giveChain } = orders[0].give;
        switch (getEngineByChainId(giveChain)) {
            case ChainEngine.Solana: {
                if (orders.find((o) => o.take.chainId !== takeChain ||
                    o.give.chainId !== giveChain ||
                    !buffersAreEqual(o.give.tokenAddress, giveToken)) !== undefined) {
                    throw new ClientError(ClientErrorType.WrongBatchOrders, `Orders with same giveChain, takeChain and giveToken are required for batch unlock`);
                }
                break;
            }
            case ChainEngine.EVM: {
                if (orders.find((o) => o.take.chainId !== takeChain || o.give.chainId !== giveChain) !== undefined) {
                    throw new ClientError(ClientErrorType.WrongBatchOrders, `Orders with same giveChain and takeChain are required for batch unlock`);
                }
                break;
            }
        }
        const address = this.getContractAddress(takeChain, ServiceType.Destination);
        const web3 = this.getConnection(takeChain);
        switch (getEngineByChainId(giveChain)) {
            case ChainEngine.Solana: {
                if (payload.solanaInitWalletReward === undefined || payload.solanaClaimUnlockReward === undefined)
                    throw new ClientError(ClientErrorType.RewardsRequired, `Rewards for send batch unlock to solana are required`);
                const method = this.dlnDestinationService.sendBatchSolanaUnlock(web3, address, orders, beneficiary, executionFee, payload.solanaInitWalletReward, payload.solanaClaimUnlockReward);
                const value = await this.getValueWithIncludedFeeAndRewards([payload.solanaInitWalletReward, payload.solanaClaimUnlockReward * BigInt(orders.length)], executionFee, takeChain, web3);
                return {
                    data: method.encodeABI(),
                    value: value.toString(),
                    to: address,
                };
            }
            case ChainEngine.EVM: {
                const method = this.dlnDestinationService.sendBatchEvmUnlock(web3, address, orders.map((o) => o.orderId), helpers.bufferToHex(beneficiary), executionFee.toString());
                const value = await this.getValueWithIncludedFee(executionFee, takeChain, web3);
                return {
                    data: method.encodeABI(),
                    value: value.toString(),
                    to: address,
                };
            }
        }
    }
    getOrderFromTransaction(params, payload) {
        return getOrderFromEvmTransaction(params.txHash, this.getConnection(params.giveChain));
    }
    async getFulfillCost(params, payload) {
        const { takeChain } = params;
        if (!Object.values(SupportedEVMChains).includes(takeChain))
            throw new ClientError(ClientErrorType.UnsupportedChain);
        const takeProvider = this.getConnection(takeChain);
        const result = await this.dlnDestinationService.getRawFulfillNativeCost(takeChain, takeProvider, payload.orderEstimationStage, payload.fulfillGasLimit, payload.fulfillCappedGasPrice, !buffersAreEqual(params.equivalentToken, params.orderTakeToken), { loggerInstance: params.loggerInstance || this.defaultLogger });
        return result;
    }
    async getClaimExecutionFee(params, payload) {
        const { action, giveChain, giveNativePrice, takeNativePrice, takeDecimals, batchSize } = params;
        const giveConnection = this.getConnection(giveChain);
        let originNativeCost;
        switch (action) {
            case "ClaimCancel": {
                originNativeCost = await this.dlnSourceService.getClaimCancelExecutionFee(giveChain, giveConnection, payload.orderEstimationStage, { ...payload, loggerInstance: params.loggerInstance });
                break;
            }
            case "ClaimUnlock": {
                originNativeCost = await this.dlnSourceService.getClaimUnlockExecutionFee(giveChain, giveConnection, payload.orderEstimationStage, { ...payload, loggerInstance: params.loggerInstance });
                break;
            }
            case "ClaimBatchUnlock": {
                if (!batchSize)
                    throw new ClientError(ClientErrorType.WrongBatchSize, "batch size required for batch unlock!");
                originNativeCost = await this.dlnSourceService.getClaimBatchUnlockExecutionFee(batchSize, giveChain, giveConnection, payload.orderEstimationStage, { ...payload, loggerInstance: params.loggerInstance });
                break;
            }
        }
        if (params.loggerInstance)
            params.loggerInstance.verbose(`originNativeCost: ${originNativeCost.totalCost}, give native price: ${giveNativePrice}, take native price: ${takeNativePrice}, take decimals: ${takeDecimals}`);
        const takeNativeCost = calcBigIntAmountViaCrossRate(originNativeCost.totalCost, giveNativePrice, this.nativeCurrencyDecimals, takeNativePrice, takeDecimals);
        return {
            claimCost: takeNativeCost,
            rewards: [],
            total: takeNativeCost,
        };
    }
    async getClaimBatchUnlockCostForSingleOrder(params, payload) {
        const originNativeCost = await this.dlnSourceService.getClaimBatchUnlockNativeCostForSingleOrder(params.batchUnlockCoverage, params.giveChain, this.getConnection(params.giveChain), payload.orderEstimationStage, { ...payload, loggerInstance: params.loggerInstance });
        const takeNativeCost = calcBigIntAmountViaCrossRate(originNativeCost.totalCost, params.giveNativePrice, this.nativeCurrencyDecimals, params.takeNativePrice, params.takeDecimals);
        return takeNativeCost;
    }
    async getSendUnlockCostForSingleOrder(params, payload) {
        const logger = (...args) => params.loggerInstance?.verbose("[getSendUnlockCostForSingleOrder]", ...args);
        const { giveChain, takeChain } = params;
        const takeConnection = this.getConnection(takeChain);
        const { totalCost: sendUnlockCostForSingleOrder } = await this.dlnDestinationService.getSendBatchUnlockNativeCostForSingleOrder(params.batchUnlockCoverage, giveChain, takeChain, takeConnection, payload.orderEstimationStage, { ...payload, loggerInstance: params.loggerInstance });
        logger("raw send_unlock cost for single order: ", sendUnlockCostForSingleOrder);
        const accountedExecutionFeeForSingleOrder = await this.getAmountWithAccountedGateTransferFee(takeChain, params.executionFeeForSingleOrder, takeConnection);
        logger("accounted execution fee for single order: ", accountedExecutionFeeForSingleOrder);
        const fixFee = await this.getGlobalFixedNativeFee(takeChain, takeConnection);
        const fixFeeForSingleOrder = (fixFee * BigInt(Math.round(params.batchUnlockCoverage * 100))) / 100n;
        logger("fix fee for single order", fixFeeForSingleOrder);
        if (!(0 <= params.discountPercents.batchUnlockFixFee && params.discountPercents.batchUnlockFixFee <= 100)) {
            throw new Error(`discountPercents.batchUnlockFixFee should be between 0 and 100`);
        }
        const fixFeeForSingleOrderWithDiscount = ((100n - params.discountPercents.batchUnlockFixFee) * fixFeeForSingleOrder) / 100n;
        logger("fix fee for single order with discount", fixFeeForSingleOrderWithDiscount);
        const totalCost = sendUnlockCostForSingleOrder + accountedExecutionFeeForSingleOrder + fixFeeForSingleOrderWithDiscount;
        logger("total cost", totalCost);
        return totalCost;
    }
    async getSendUnlockCost(params, payload) {
        let logger = (...args) => { };
        if (params?.loggerInstance) {
            logger = (...args) => params.loggerInstance.verbose("[getSendUnlockCost]", ...args);
        }
        const { giveChain, takeChain } = params;
        const takeConnection = this.getConnection(takeChain);
        const nativeCost = await this.dlnDestinationService.getSendBatchUnlockNativeCost(params.batchSize, giveChain, takeChain, takeConnection, payload.orderEstimationStage, { ...payload, loggerInstance: params.loggerInstance });
        logger("native cost for send unlock", nativeCost);
        const accountedExecutionFee = await this.getAmountWithAccountedGateTransferFee(takeChain, params.executionFee, takeConnection);
        logger("accounted execution fee", accountedExecutionFee);
        const fixFee = await this.getGlobalFixedNativeFee(takeChain, takeConnection);
        logger("fix fee", fixFee);
        const result = nativeCost.totalCost + accountedExecutionFee + fixFee;
        logger("final cost", result);
        return {
            transactionCost: nativeCost,
            accountedExecutionFee,
            fixFee,
            totalCost: result,
        };
    }
    async getSendCancelCost(params, payload) {
        let logger = (...args) => { };
        if (params?.loggerInstance) {
            logger = (...args) => params.loggerInstance.verbose("[getSendCancelCost]", ...args);
        }
        const { giveChain, takeChain } = params;
        const takeConnection = this.getConnection(takeChain);
        const nativeCost = await this.dlnDestinationService.getSendCancelCost(takeChain, giveChain, takeConnection, payload.orderEstimationStage, { loggerInstance: params.loggerInstance });
        logger("native cost for send cancel", nativeCost);
        const accountedExecutionFee = await this.getAmountWithAccountedGateTransferFee(takeChain, params.executionFee, takeConnection);
        logger("accounted execution fee", accountedExecutionFee);
        const fixFee = await this.getGlobalFixedNativeFee(takeChain, takeConnection);
        logger("fix fee", fixFee);
        const result = nativeCost.totalCost + accountedExecutionFee + fixFee;
        logger("final cost", result);
        return {
            transactionCost: nativeCost,
            accountedExecutionFee,
            fixFee,
            totalCost: result,
        };
    }
    getContractAddress(chainId, type) {
        const chainConfig = this.config.chainConfig[chainId];
        if (chainConfig === undefined)
            throw new ClientError(ClientErrorType.UnsupportedChain, `Config for chain ${ChainId[chainId]} is missing`);
        const typeMap = {
            [ServiceType.Source]: "dlnSourceAddress",
            [ServiceType.Destination]: "dlnDestinationAddress",
            [ServiceType.DebridgeGate]: "deBridgeGateAddress",
            [ServiceType.CrosschainForwarder]: "crossChainForwarderAddress",
        };
        const address = chainConfig[typeMap[type]]; // can't set address to undefined in config
        return address;
    }
}
