// @debridge-finance/solana-utils/dist/esm/constants.js
export const BPS_DENOMINATOR = 10000;
export const SOLANA_CHAIN_ID = 7565164;
export const MAX_TX_SIZE = 1232;
export const FAKE_BLOCKHASH = "1".repeat(32);
export const DEPLOY_INFO_PREFIX = 2;
export const SUBMISSION_PREFIX = 1;
export const deBridge = {
    Seeds: {
        STATE: "DEBRIDGE_STATE",
        BRIDGE: "BRIDGE",
        WRAPPER_TOKEN: "WRAPPER_TOKEN",
        SEND_BRIDGE_MAP: "WRAPPER_TOKEN",
        BRIDGE_MINTER: "BRIDGE_MINTER",
        CHAIN_SUPPORT_INFO: "CHAIN_SUPPORT_INFO",
        BRIDGE_FEE_INFO: "BRIDGE_FEE_INFO",
        DISCOUNT_INFO: "DISCOUNT_INFO",
        ORACLE_KEYS: "SIGNATURES",
        EXTERNAL_CALL_STORAGE: "EXTERNAL_CALL_STORAGE",
        EXTERNAL_CALL_META: "EXTERNAL_CALL_META",
        NONCE: "NONCE",
        SUBMISSION: "SUBMISSION",
        SUBMISSION_AUTH: "SUBMISSION_AUTH",
        NO_DISCOUNT: "NO_DISCOUNT",
        CLAIM_MARKER: "CLAIM_MARKER",
    },
    Flags: {
        UNWRAP_ETH: 1,
        REVERT_IF_EXTERNAL_FAIL: 2,
        PROXY_WITH_SENDER: 4,
        SEND_HASHED_DATA: 8,
        SEND_TO_ATA: Math.pow(2, 31),
    },
};
export const deBridgeSettings = {
    Seeds: {
        TOKEN_METADATA: "TOKEN_METADATA",
        WRAPPER_TOKEN: "WRAPPER_TOKEN",
        STATE: "DEBRIDGE_STATE",
        CONFIRMATION_STORAGE: "SIGNATURES",
        BRIDGE_FEE_INFO: "BRIDGE_FEE_INFO",
        CHAIN_SUPPORT_INFO: "CHAIN_SUPPORT_INFO",
        DISCOUNT_INFO: "DISCOUNT_INFO",
        BRIDGE: "BRIDGE",
        NO_BRIDGE_FEE_INFO: "DEFAULT_BRIDGE_FEE_INFO",
        NO_DISCOUNT: "NO_DISCOUNT",
        SEND_BRIDGE_MAP: "WRAPPER_TOKEN",
    },
};
export const dlnSrc = {
    Seeds: {
        STATE: "STATE",
        NONCE: "NONCE",
        ORDER_STATE: "GIVE_ORDER_STATE",
        AUTHORIZED_NATIVE_SENDER: "AUTHORIZED_NATIVE_SENDER",
        GIVE_ORDER_WALLET: "GIVE_ORDER_WALLET",
        FEE_LEDGER_WALLET: "FEE_LEDGER_WALLET",
        FEE_LEDGER: "FEE LEDGER",
    },
};
export const dlnDst = {
    Seeds: {
        AUTHORIZED_SRC_CONTRACT: "AUTHORIZED_SRC_CONTRACT",
        AUTHORIZED_DST_NATIVE_SENDER: "AUTHORIZED_DST_NATIVE_SENDER",
        ORDER_STATE: "TAKE_ORDER_STATE",
        TAKE_PATCH: "TAKE_ORDER_PATCH",
        STATE: "STATE",
    },
};
