// @debridge-finance/dln-client/dist/cjs/swapConnector/bitrcock-usdb-lp.connector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BitrockUSDBLP = exports.BitrockBuffers = exports.BitrockAddresses = void 0;
const common_types_1 = require("../common.types");
const utils_1 = require("../utils");
const uniswapv2_connector_1 = require("./uniswapv2.connector");
exports.BitrockAddresses = {
    UsdbAddress: "0xD647a6E1F898c39bA594443c19De0872EA935310",
    wBrockAddress: "0x413f0E3A440abA7A15137F4278121450416882d5",
    RockAddress: "0x826628622a8B55F4AF6aC2A1DD1273bF837e651a",
    RockRouterAddress: "0xeeabd314e2eE640B1aca3B27808972B05c7f6A3b",
};
exports.BitrockBuffers = {
    UsdbAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, exports.BitrockAddresses.UsdbAddress),
    wBrockAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, exports.BitrockAddresses.wBrockAddress),
    RockAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, exports.BitrockAddresses.RockAddress),
    BrockAddress: Buffer.alloc(20, 0),
    RockRouterAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, exports.BitrockAddresses.RockRouterAddress),
};
class BitrockUSDBLP extends uniswapv2_connector_1.UniswapV2Connector {
    constructor(connection) {
        super(connection, common_types_1.ChainId.Bitrock, [
            // USDB/wBROCK
            [exports.BitrockBuffers.UsdbAddress, exports.BitrockBuffers.wBrockAddress],
            // ROCK/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0x826628622a8b55f4af6ac2a1dd1273bf837e651a"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // MSMIL/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0xaa72d86210ac33bca2de6139403f9af37398e721"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // BRW/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0xa7c9bb7a6fd13dca85e9df12b4f91dc8bede225e"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // APE/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0xe8a4f717ac5b08bccc7240d649af653b2577b36a"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // CHARM/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0xa056871e6796315c558280bff7e7f5d2c5b1f6fb"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // MICKEY/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0x978adb9fbf147c1316b3c04230c88b356f3f4513"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // RR/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0xe4bda992ecad56986704e21f1622e5f8aceaddb1"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // HE-MAN/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0x01c8b36c965adb150cd4244edb463d4ba96549d4"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // DUMBASS/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0x47094defb05e50309499acd572fa107c89c3b305"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // BDOG/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0xa9fae95179ba375b0e7e6bcb4932fb484138c6e4"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // jamma/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0xfaF9a02dB868cb557aA9EF6C9eEAac34aFD013e2"),
                exports.BitrockBuffers.wBrockAddress,
            ],
            // SNIPER/wBROCK
            [
                (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Bitrock, "0x8a6162CdcBa0878212BD5C6857ff977E218E7772"),
                exports.BitrockBuffers.wBrockAddress,
            ],
        ], exports.BitrockBuffers.RockRouterAddress, exports.BitrockBuffers.wBrockAddress);
    }
}
exports.BitrockUSDBLP = BitrockUSDBLP;
