// @debridge-finance/dln-client/dist/cjs/swapConnector/neon-usdc-lp.connector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NeonICS = exports.NeonMoraswap = exports.NeonUsdcLp = exports.NeonTokenAddressBuffers = exports.NeonTokenAddresses = void 0;
const common_types_1 = require("../common.types");
const utils_1 = require("../utils");
const uniswapv2_connector_1 = require("./uniswapv2.connector");
exports.NeonTokenAddresses = {
    UsdcAddress: "0xEA6B04272f9f62F997F666F07D3a974134f7FFb9",
    UsdtAddress: "0x5f0155d08ef4aae2b500aefb64a3419da8bb611a",
    wSolAddress: "0x5f38248f339Bf4e84A2caf4e4c0552862dC9F82a",
    wNeonAddress: "0x202C35e517Fa803B537565c40F0a6965D7204609",
    MoraSwapV2RouterAddress: "0x594e37B9F39f5D31DEc4a8c1cC4fe2E254153034",
    ICSV2RouterAddress: "0xBb5e1777A331ED93E07cF043363e48d320eb96c4",
};
exports.NeonTokenAddressBuffers = {
    UsdcAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Neon, exports.NeonTokenAddresses.UsdcAddress),
    UsdtAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Neon, exports.NeonTokenAddresses.UsdtAddress),
    wSolAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Neon, exports.NeonTokenAddresses.wSolAddress),
    wNeonAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Neon, exports.NeonTokenAddresses.wNeonAddress),
    NeonAddress: Buffer.alloc(20, 0),
    MoraSwapV2RouterAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Neon, exports.NeonTokenAddresses.MoraSwapV2RouterAddress),
    ICSV2RouterAddress: (0, utils_1.tokenStringToBuffer)(common_types_1.ChainId.Neon, exports.NeonTokenAddresses.ICSV2RouterAddress),
};
class NeonUsdcLp extends uniswapv2_connector_1.UniswapV2Connector {
    constructor(connection) {
        super(connection, common_types_1.ChainId.Neon, [
            [exports.NeonTokenAddressBuffers.UsdcAddress, exports.NeonTokenAddressBuffers.wSolAddress],
            [exports.NeonTokenAddressBuffers.wSolAddress, exports.NeonTokenAddressBuffers.wNeonAddress],
        ], exports.NeonTokenAddressBuffers.MoraSwapV2RouterAddress, exports.NeonTokenAddressBuffers.wNeonAddress);
    }
}
exports.NeonUsdcLp = NeonUsdcLp;
class NeonMoraswap extends NeonUsdcLp {
}
exports.NeonMoraswap = NeonMoraswap;
class NeonICS extends uniswapv2_connector_1.UniswapV2Connector {
    constructor(connection) {
        super(connection, common_types_1.ChainId.Neon, [
            // USDC/USDT
            [exports.NeonTokenAddressBuffers.UsdcAddress, exports.NeonTokenAddressBuffers.UsdtAddress],
            // USDT/wSOL
            [exports.NeonTokenAddressBuffers.wSolAddress, exports.NeonTokenAddressBuffers.UsdtAddress],
            // USDT/wNEON
            [exports.NeonTokenAddressBuffers.wNeonAddress, exports.NeonTokenAddressBuffers.UsdtAddress],
        ], exports.NeonTokenAddressBuffers.ICSV2RouterAddress, exports.NeonTokenAddressBuffers.wNeonAddress);
    }
}
exports.NeonICS = NeonICS;
