// @debridge-finance/dln-client/dist/cjs/swapConnector/openocean.connector.js
"use strict";
var _OpenOceanSwapConnector_instances, _OpenOceanSwapConnector_apiKey, _OpenOceanSwapConnector_getDecimalsLambda, _OpenOceanSwapConnector_getGasPriceLambda, _OpenOceanSwapConnector_disabledDexIds, _OpenOceanSwapConnector_enabledDexIds, _OpenOceanSwapConnector_maskCredentialFunction, _OpenOceanSwapConnector_basicUrl, _OpenOceanSwapConnector_timeoutInMs, _OpenOceanSwapConnector_referrer, _OpenOceanSwapConnector_callOpenOcean, _OpenOceanSwapConnector_mapChainId, _OpenOceanSwapConnector_mapAddress, _OpenOceanSwapConnector_deleteAmountDecimals, _OpenOceanSwapConnector_getGasPrice;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenOceanSwapConnector = exports.openOceanIaaSChainMapper = exports.openOceanNativeTokens = void 0;
const tslib_1 = require("tslib");
const solana_utils_1 = require("@debridge-finance/solana-utils");
const bignumber_js_1 = tslib_1.__importDefault(require("bignumber.js"));
const web3_1 = tslib_1.__importDefault(require("web3"));
const common_types_1 = require("../common.types");
const swap_connector_1 = require("./swap.connector");
const swap_connector_error_type_1 = require("./errors/swap-connector-error.type");
const swap_connector_error_1 = require("./errors/swap-connector.error");
const swap_connector_logger_1 = require("./swap-connector.logger");
var OpenOceanSupportedChainId;
(function (OpenOceanSupportedChainId) {
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Ethereum"] = 1] = "Ethereum";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["BSC"] = 56] = "BSC";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Arbitrum"] = 42161] = "Arbitrum";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Avalanche"] = 43114] = "Avalanche";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Optimism"] = 10] = "Optimism";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Polygon"] = 137] = "Polygon";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Base"] = 8453] = "Base";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Linea"] = 59144] = "Linea";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Gnosis"] = 100000002] = "Gnosis";
    OpenOceanSupportedChainId[OpenOceanSupportedChainId["Metis"] = 100000004] = "Metis";
})(OpenOceanSupportedChainId || (OpenOceanSupportedChainId = {}));
exports.openOceanNativeTokens = {
    [OpenOceanSupportedChainId.Ethereum]: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
    [OpenOceanSupportedChainId.BSC]: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
    [OpenOceanSupportedChainId.Arbitrum]: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
    [OpenOceanSupportedChainId.Avalanche]: "0x0000000000000000000000000000000000000000",
    [OpenOceanSupportedChainId.Optimism]: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
    [OpenOceanSupportedChainId.Polygon]: "0x0000000000000000000000000000000000001010",
    [OpenOceanSupportedChainId.Base]: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
    [OpenOceanSupportedChainId.Linea]: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
    [OpenOceanSupportedChainId.Gnosis]: "0x0000000000000000000000000000000000000000",
    [OpenOceanSupportedChainId.Metis]: "0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000",
};
exports.openOceanIaaSChainMapper = {
    [OpenOceanSupportedChainId.Gnosis]: 100,
    [OpenOceanSupportedChainId.Metis]: 1088,
};
class OpenOceanSwapConnector extends swap_connector_1.SwapConnector {
    constructor(config) {
        super();
        _OpenOceanSwapConnector_instances.add(this);
        _OpenOceanSwapConnector_apiKey.set(this, void 0);
        _OpenOceanSwapConnector_getDecimalsLambda.set(this, void 0);
        _OpenOceanSwapConnector_getGasPriceLambda.set(this, void 0);
        _OpenOceanSwapConnector_disabledDexIds.set(this, void 0);
        _OpenOceanSwapConnector_enabledDexIds.set(this, void 0);
        _OpenOceanSwapConnector_maskCredentialFunction.set(this, void 0);
        _OpenOceanSwapConnector_basicUrl.set(this, "https://open-api-pro.openocean.finance");
        _OpenOceanSwapConnector_timeoutInMs.set(this, void 0);
        _OpenOceanSwapConnector_referrer.set(this, void 0);
        tslib_1.__classPrivateFieldSet(this, _OpenOceanSwapConnector_apiKey, config.apiKey, "f");
        tslib_1.__classPrivateFieldSet(this, _OpenOceanSwapConnector_disabledDexIds, config.disabledDexIds || [], "f");
        tslib_1.__classPrivateFieldSet(this, _OpenOceanSwapConnector_enabledDexIds, config.enabledDexIds || [], "f");
        tslib_1.__classPrivateFieldSet(this, _OpenOceanSwapConnector_maskCredentialFunction, (text) => text.replaceAll(config.apiKey, "<API KEY>"), "f");
        tslib_1.__classPrivateFieldSet(this, _OpenOceanSwapConnector_timeoutInMs, config.timeoutInMs, "f");
        tslib_1.__classPrivateFieldSet(this, _OpenOceanSwapConnector_getDecimalsLambda, config.getDecimalsLambda, "f");
        tslib_1.__classPrivateFieldSet(this, _OpenOceanSwapConnector_getGasPriceLambda, config.getGasPriceLambda, "f");
        tslib_1.__classPrivateFieldSet(this, _OpenOceanSwapConnector_referrer, config.referrer, "f");
    }
    getSupportedChains() {
        return Object.values(OpenOceanSupportedChainId)
            .filter((v) => typeof v === "number")
            .map((_) => Number(_));
    }
    async getEstimate(request, context) {
        const logger = new swap_connector_logger_1.SwapConnectorLogger(context.logger, OpenOceanSwapConnector.name, tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_maskCredentialFunction, "f"));
        if (!this.getSupportedChains().includes(request.chainId)) {
            const message = `Chain ${common_types_1.ChainId[request.chainId]} is not supported`;
            throw new swap_connector_error_1.SwapConnectorError(OpenOceanSwapConnector.name, swap_connector_error_type_1.SwapConnectorErrorType.UnsupportedChain, message);
        }
        const chainId = request.chainId;
        const requestBody = {
            inTokenAddress: tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_mapAddress).call(this, chainId, request.fromTokenAddress),
            outTokenAddress: tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_mapAddress).call(this, chainId, request.toTokenAddress),
            amount: await tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_deleteAmountDecimals).call(this, chainId, request.fromTokenAddress, request.amountIn),
            gasPrice: (await tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_getGasPrice).call(this, chainId)).toString(),
        };
        const params = new URLSearchParams(requestBody);
        if (tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_disabledDexIds, "f").length) {
            params.append("disabledDexIds", tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_disabledDexIds, "f").join(","));
        }
        if (tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_enabledDexIds, "f").length) {
            params.append("enabledDexIds", tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_enabledDexIds, "f").join(","));
        }
        const response = await tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_callOpenOcean).call(this, `/v4/${tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_mapChainId).call(this, request.chainId)}/quote`, params, logger);
        return {
            amountIn: BigInt(response.data.inAmount),
            amountOut: BigInt(response.data.outAmount),
            chainId: request.chainId,
            fromTokenAddress: request.fromTokenAddress,
            toTokenAddress: request.toTokenAddress,
            rawConnectorResponse: response,
        };
    }
    async getSwap(request, context) {
        if (!this.getSupportedChains().includes(request.chainId)) {
            const message = `Chain ${common_types_1.ChainId[request.chainId]} is not supported`;
            throw new swap_connector_error_1.SwapConnectorError(OpenOceanSwapConnector.name, swap_connector_error_type_1.SwapConnectorErrorType.UnsupportedChain, message);
        }
        const logger = new swap_connector_logger_1.SwapConnectorLogger(context.logger, OpenOceanSwapConnector.name, tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_maskCredentialFunction, "f"));
        const chainId = request.chainId;
        const requestBody = {
            inTokenAddress: tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_mapAddress).call(this, chainId, request.fromTokenAddress),
            outTokenAddress: tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_mapAddress).call(this, chainId, request.toTokenAddress),
            amount: await tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_deleteAmountDecimals).call(this, chainId, request.fromTokenAddress, request.amountIn),
            gasPrice: (await tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_getGasPrice).call(this, chainId)).toString(),
            slippage: (0, swap_connector_1.bpsToPercent)(request.slippageBps).toString(),
            account: solana_utils_1.helpers.bufferToHex(request.fromAddress),
        };
        const params = new URLSearchParams(requestBody);
        if (tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_disabledDexIds, "f").length) {
            params.append("disabledDexIds", tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_disabledDexIds, "f").join(","));
        }
        if (tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_enabledDexIds, "f").length) {
            params.append("enabledDexIds", tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_enabledDexIds, "f").join(","));
        }
        if (tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_referrer, "f")) {
            params.append("referrer", tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_referrer, "f"));
        }
        const response = await tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_callOpenOcean).call(this, `/v4/${tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_instances, "m", _OpenOceanSwapConnector_mapChainId).call(this, request.chainId)}/swap`, params, logger);
        return {
            amountIn: BigInt(response.data.inAmount),
            amountOut: BigInt(response.data.outAmount),
            engine: "evm",
            tx: {
                data: response.data.data,
                to: response.data.to,
                value: response.data.value,
            },
            fromToken: request.fromTokenAddress,
            toToken: request.toTokenAddress,
            rawConnectorResponse: response,
        };
    }
    get connectorName() {
        return "OpenOcean";
    }
}
exports.OpenOceanSwapConnector = OpenOceanSwapConnector;
_OpenOceanSwapConnector_apiKey = new WeakMap(), _OpenOceanSwapConnector_getDecimalsLambda = new WeakMap(), _OpenOceanSwapConnector_getGasPriceLambda = new WeakMap(), _OpenOceanSwapConnector_disabledDexIds = new WeakMap(), _OpenOceanSwapConnector_enabledDexIds = new WeakMap(), _OpenOceanSwapConnector_maskCredentialFunction = new WeakMap(), _OpenOceanSwapConnector_basicUrl = new WeakMap(), _OpenOceanSwapConnector_timeoutInMs = new WeakMap(), _OpenOceanSwapConnector_referrer = new WeakMap(), _OpenOceanSwapConnector_instances = new WeakSet(), _OpenOceanSwapConnector_callOpenOcean = async function _OpenOceanSwapConnector_callOpenOcean(path, query, logger) {
    try {
        return await this.call({
            basicUrl: tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_basicUrl, "f"),
            path,
            query,
            headers: {
                apikey: tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_apiKey, "f"),
            },
            timeoutInMs: tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_timeoutInMs, "f"),
        }, {
            logger,
        });
    }
    catch (e) {
        const error = e;
        throw new swap_connector_error_1.SwapConnectorError(OpenOceanSwapConnector.name, swap_connector_error_type_1.SwapConnectorErrorType.InternalError, error.message);
    }
}, _OpenOceanSwapConnector_mapChainId = function _OpenOceanSwapConnector_mapChainId(chainId) {
    return exports.openOceanIaaSChainMapper[chainId] || chainId;
}, _OpenOceanSwapConnector_mapAddress = function _OpenOceanSwapConnector_mapAddress(chainId, address) {
    let stringified = solana_utils_1.helpers.bufferToHex(address);
    if (stringified === "0x0000000000000000000000000000000000000000") {
        const nativeToken = exports.openOceanNativeTokens[chainId];
        if (!nativeToken)
            throw new swap_connector_error_1.SwapConnectorError(OpenOceanSwapConnector.name, swap_connector_error_type_1.SwapConnectorErrorType.InternalError, `Not mapped native token`);
        stringified = nativeToken;
    }
    return stringified;
}, _OpenOceanSwapConnector_deleteAmountDecimals = async function _OpenOceanSwapConnector_deleteAmountDecimals(chainId, tokenAddress, amount) {
    const decimals = await tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_getDecimalsLambda, "f").call(this, chainId, tokenAddress);
    return new bignumber_js_1.default(amount.toString()).div((10 ** decimals).toString()).toString();
}, _OpenOceanSwapConnector_getGasPrice = async function _OpenOceanSwapConnector_getGasPrice(chainId) {
    const gasPrice = await tslib_1.__classPrivateFieldGet(this, _OpenOceanSwapConnector_getGasPriceLambda, "f").call(this, chainId);
    return web3_1.default.utils.fromWei(gasPrice, "gwei");
};
