// @debridge-finance/dln-client/dist/esm/evm/core/utils/get.evm.info.js
import { ChainId, ZERO_EVM_ADDRESS } from "../../../common.types";
const nativeTokenInfo = {
    [ChainId.Polygon]: {
        name: "Polygon",
        symbol: "MATIC",
    },
    [ChainId.BSC]: {
        name: "BNB",
        symbol: "BNB",
    },
    [ChainId.Ethereum]: {
        name: "Ethereum",
        symbol: "ETH",
    },
    [ChainId.Arbitrum]: {
        name: "Ethereum",
        symbol: "ETH",
    },
    [ChainId.Avalanche]: {
        name: "Avalanche",
        symbol: "AVAX",
    },
    [ChainId.Fantom]: {
        name: "Fantom",
        symbol: "FTM",
    },
    [ChainId.Linea]: {
        name: "Ethereum",
        symbol: "ETH",
    },
    [ChainId.Optimism]: {
        name: "Ethereum",
        symbol: "ETH",
    },
    [ChainId.Base]: {
        name: "Ethereum",
        symbol: "ETH",
    },
    [ChainId.Neon]: {
        name: "Neon",
        symbol: "NEON",
    },
    [ChainId.Gnosis]: {
        name: "xDAI",
        symbol: "XDAI",
    },
    [ChainId.LightLink]: {
        name: "Ethereum",
        symbol: "ETH",
    },
    [ChainId.Metis]: {
        name: "Metis",
        symbol: "METIS",
    },
    [ChainId.Bitrock]: {
        name: "Bitrock",
        symbol: "BROCK",
    },
};
const IERC20Abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "balanceOf",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "decimals",
        outputs: [
            {
                internalType: "uint8",
                name: "",
                type: "uint8",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "name",
        outputs: [
            {
                internalType: "string",
                name: "",
                type: "string",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "symbol",
        outputs: [
            {
                internalType: "string",
                name: "",
                type: "string",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
//#endregion
export async function getEvmDecimals(address, web3) {
    if (address === ZERO_EVM_ADDRESS) {
        return 18;
    }
    const contract = new web3.eth.Contract(IERC20Abi, address);
    const decimals = await contract.methods.decimals().call();
    return parseInt(decimals);
}
export const getEvmAccountBalance = async (tokenAddress, address, web3) => {
    if (tokenAddress === ZERO_EVM_ADDRESS) {
        return BigInt(await web3.eth.getBalance(address));
    }
    const contract = new web3.eth.Contract(IERC20Abi, tokenAddress);
    return BigInt((await contract.methods.balanceOf(address).call()));
};
export async function getEvmTokenName(address, chainId, web3) {
    if (address === ZERO_EVM_ADDRESS) {
        return nativeTokenInfo[chainId].name;
    }
    const contract = new web3.eth.Contract(IERC20Abi, address);
    const result = await contract.methods.name().call();
    return result;
}
export async function getEvmTokenSymbol(address, chainId, web3) {
    if (address === ZERO_EVM_ADDRESS) {
        return nativeTokenInfo[chainId].symbol;
    }
    const contract = new web3.eth.Contract(IERC20Abi, address);
    const result = await contract.methods.symbol().call();
    return result;
}
