// @debridge-finance/dln-client/dist/esm/evm/core/utils/order.data.to.evm.js
import { helpers } from "@debridge-finance/solana-utils";
import { Order } from "../../../order";
function formatExternalCall(externalCall, useHash) {
    if (externalCall) {
        return helpers.bufferToHex(useHash ? Order.getExternalCallHash(externalCall) : externalCall.externalCallData);
    }
    else
        return "0x";
}
export function orderToEvm(order, useExtCallHash = false) {
    if ("nonce" in order && order.nonce !== undefined) {
        return {
            makerOrderNonce: order.nonce.toString(),
            makerSrc: order.maker,
            giveAmount: String(order.give.amount),
            giveChainId: Number(order.give.chainId),
            giveTokenAddress: helpers.bufferToHex(Buffer.from(order.give.tokenAddress)),
            takeAmount: String(order.take.amount),
            takeChainId: Number(order.take.chainId),
            takeTokenAddress: helpers.bufferToHex(Buffer.from(order.take.tokenAddress)),
            receiverDst: order.receiver,
            allowedCancelBeneficiarySrc: order.allowedCancelBeneficiary || "0x",
            allowedTakerDst: order.allowedTaker || "0x",
            externalCall: formatExternalCall(order.externalCall, useExtCallHash),
            givePatchAuthoritySrc: helpers.bufferToHex(Buffer.from(order.givePatchAuthority)),
            orderAuthorityAddressDst: order.orderAuthorityDstAddress,
        };
    }
    return {
        giveAmount: String(order.give.amount),
        giveChainId: Number(order.give.chainId),
        giveTokenAddress: helpers.bufferToHex(Buffer.from(order.give.tokenAddress)),
        takeAmount: String(order.take.amount),
        takeChainId: Number(order.take.chainId),
        takeTokenAddress: helpers.bufferToHex(Buffer.from(order.take.tokenAddress)),
        receiverDst: order.receiver,
        allowedCancelBeneficiarySrc: order.allowedCancelBeneficiary || "0x",
        allowedTakerDst: order.allowedTaker || "0x",
        externalCall: formatExternalCall(order.externalCall, useExtCallHash),
        givePatchAuthoritySrc: helpers.bufferToHex(Buffer.from(order.givePatchAuthority)),
        orderAuthorityAddressDst: order.orderAuthorityDstAddress,
    };
}
