// @debridge-finance/dln-client/dist/cjs/evm/core/types/common.type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedEVMChains = void 0;
var SupportedEVMChains;
(function (SupportedEVMChains) {
    SupportedEVMChains[SupportedEVMChains["Arbitrum"] = 42161] = "Arbitrum";
    SupportedEVMChains[SupportedEVMChains["Avalanche"] = 43114] = "Avalanche";
    SupportedEVMChains[SupportedEVMChains["Base"] = 8453] = "Base";
    SupportedEVMChains[SupportedEVMChains["Bitrock"] = 100000005] = "Bitrock";
    SupportedEVMChains[SupportedEVMChains["BSC"] = 56] = "BSC";
    SupportedEVMChains[SupportedEVMChains["Ethereum"] = 1] = "Ethereum";
    SupportedEVMChains[SupportedEVMChains["Fantom"] = 250] = "Fantom";
    SupportedEVMChains[SupportedEVMChains["Gnosis"] = 100000002] = "Gnosis";
    SupportedEVMChains[SupportedEVMChains["LightLink"] = 100000003] = "LightLink";
    SupportedEVMChains[SupportedEVMChains["Linea"] = 59144] = "Linea";
    SupportedEVMChains[SupportedEVMChains["Metis"] = 100000004] = "Metis";
    SupportedEVMChains[SupportedEVMChains["Neon"] = 100000001] = "Neon";
    SupportedEVMChains[SupportedEVMChains["Optimism"] = 10] = "Optimism";
    SupportedEVMChains[SupportedEVMChains["Polygon"] = 137] = "Polygon";
})(SupportedEVMChains || (exports.SupportedEVMChains = SupportedEVMChains = {}));
