// @debridge-finance/dln-client/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setPredictGasPriceCachedChains = exports.setPredictGasPriceNodeCache = exports.Env = exports.CommonDlnClient = exports.formatAmount = exports.fixDecimals = exports.tokenStringToBuffer = exports.tokenAddressToString = exports.getNativeToken = exports.buffersAreEqual = exports.Evm = exports.Solana = void 0;
const tslib_1 = require("tslib");
exports.Solana = tslib_1.__importStar(require("./solana/index"));
exports.Evm = tslib_1.__importStar(require("./evm/index"));
tslib_1.__exportStar(require("./order"), exports);
tslib_1.__exportStar(require("./common.types"), exports);
tslib_1.__exportStar(require("./swapConnector"), exports);
tslib_1.__exportStar(require("./priceFeeds"), exports);
tslib_1.__exportStar(require("./errors/client-error.error"), exports);
tslib_1.__exportStar(require("./errors/client-error-type.enum"), exports);
var utils_1 = require("./utils");
Object.defineProperty(exports, "buffersAreEqual", { enumerable: true, get: function () { return utils_1.buffersAreEqual; } });
Object.defineProperty(exports, "getNativeToken", { enumerable: true, get: function () { return utils_1.getNativeToken; } });
Object.defineProperty(exports, "tokenAddressToString", { enumerable: true, get: function () { return utils_1.tokenAddressToString; } });
Object.defineProperty(exports, "tokenStringToBuffer", { enumerable: true, get: function () { return utils_1.tokenStringToBuffer; } });
Object.defineProperty(exports, "fixDecimals", { enumerable: true, get: function () { return utils_1.fixDecimals; } });
Object.defineProperty(exports, "formatAmount", { enumerable: true, get: function () { return utils_1.formatAmount; } });
tslib_1.__exportStar(require("./common/logger"), exports);
var client_1 = require("./common/client");
Object.defineProperty(exports, "CommonDlnClient", { enumerable: true, get: function () { return client_1.CommonDlnClient; } });
var env_1 = require("./common/env");
Object.defineProperty(exports, "Env", { enumerable: true, get: function () { return env_1.Env; } });
var gasprice_service_1 = require("./evm/core/services/gasprice.service");
Object.defineProperty(exports, "setPredictGasPriceNodeCache", { enumerable: true, get: function () { return gasprice_service_1.setPredictGasPriceNodeCache; } });
Object.defineProperty(exports, "setPredictGasPriceCachedChains", { enumerable: true, get: function () { return gasprice_service_1.setPredictGasPriceCachedChains; } });
