// @debridge-finance/dln-client/examples/initClient.ts
import { Connection } from "@solana/web3.js";
import Web3 from "web3";

import { Evm, ChainId, Solana, CommonDlnClient } from "../src";

export function initClients(): [Solana.DlnClient, Evm.DlnClient, CommonDlnClient<Evm.DlnClient | Solana.DlnClient>] {
  const evmClient = new Evm.DlnClient({
    chainConfig: {
      [ChainId.BSC]: {
        crossChainForwarderAddress: "0xc31fc94F3Fd088eE53ac915D6e8a14fF25a23C47",
        deBridgeGateAddress: "0x43dE2d77BF8027e25dBD179B491e8d64f38398aA",
        dlnDestinationAddress: "0xE7351Fd770A37282b91D153Ee690B63579D6dd7f",
        dlnSourceAddress: "0xeF4fB24aD0916217251F553c0596F8Edc630EB66",
        connection: new Web3("https://bsc-dataseed1.binance.org/"),
      },
    },
    enableContractsCache: true,
  });

  const solanaClient = new Solana.DlnClient(
    new Connection("https://api.mainnet-beta.solana.com"),
    "src5qyZHqTqecJV4aY6Cb6zDZLMDzrDKKezs22MPHr4",
    "dst5MGcFPoBeREFAA5E3tU5ij8m5uVYwkzkSAbsLbNo",
    "DEbrdGj3HsRsAzx6uH4MKyREKxVAfBydijLUF3ygsFfh",
    "DeSetTwWhjZq6Pz9Kfdo1KoS5NqtsM6G8ERbX4SSCSft",
  );

  const commonClient = new CommonDlnClient<Evm.DlnClient | Solana.DlnClient>(solanaClient, evmClient);

  return [solanaClient, evmClient, commonClient];
}
