// @debridge-finance/dln-client/dist/cjs/solana/idl/executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IDL = void 0;
exports.IDL = {
    version: "1.1.1",
    name: "extcall_program",
    instructions: [
        {
            name: "initializeState",
            accounts: [
                {
                    name: "state",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "protocolAuthority",
                    isMut: true,
                    isSigner: true,
                },
                {
                    name: "systemProgram",
                    isMut: false,
                    isSigner: false,
                },
            ],
            args: [
                {
                    name: "whitelist",
                    type: {
                        vec: "publicKey",
                    },
                },
            ],
        },
        {
            name: "updateState",
            accounts: [
                {
                    name: "state",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "protocolAuthority",
                    isMut: false,
                    isSigner: true,
                },
                {
                    name: "systemProgram",
                    isMut: false,
                    isSigner: false,
                },
            ],
            args: [
                {
                    name: "whitelist",
                    type: {
                        vec: "publicKey",
                    },
                },
            ],
        },
        {
            name: "initializeExtcallMeta",
            accounts: [
                {
                    name: "state",
                    isMut: false,
                    isSigner: false,
                },
                {
                    name: "extcallMeta",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "liquiditySourceWallet",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "liquiditySourceAuth",
                    isMut: true,
                    isSigner: true,
                },
                {
                    name: "executeWallet",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "tokenMint",
                    isMut: false,
                    isSigner: false,
                },
                {
                    name: "creator",
                    isMut: false,
                    isSigner: true,
                },
                {
                    name: "splTokenProgram",
                    isMut: false,
                    isSigner: false,
                },
                {
                    name: "systemProgram",
                    isMut: false,
                    isSigner: false,
                },
                {
                    name: "instructions",
                    isMut: false,
                    isSigner: false,
                },
            ],
            args: [
                {
                    name: "extcallId",
                    type: {
                        array: ["u8", 32],
                    },
                },
                {
                    name: "externalInstructionsHash",
                    type: {
                        array: ["u8", 32],
                    },
                },
                {
                    name: "providedAmount",
                    type: "u64",
                },
                {
                    name: "providedToken",
                    type: {
                        defined: "ProvidedToken",
                    },
                },
                {
                    name: "fallbackAuthority",
                    type: "publicKey",
                },
            ],
        },
        {
            name: "fillExtcallStorage",
            accounts: [
                {
                    name: "extcallStorage",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "owner",
                    isMut: true,
                    isSigner: true,
                },
                {
                    name: "systemProgram",
                    isMut: false,
                    isSigner: false,
                },
            ],
            args: [
                {
                    name: "extcallId",
                    type: {
                        array: ["u8", 32],
                    },
                },
                {
                    name: "space",
                    type: "u64",
                },
                {
                    name: "offset",
                    type: "u64",
                },
                {
                    name: "bytes",
                    type: "bytes",
                },
            ],
        },
        {
            name: "closeExtcall",
            accounts: [
                {
                    name: "extcallMeta",
                    isMut: false,
                    isSigner: false,
                },
                {
                    name: "extcallStorage",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "owner",
                    isMut: true,
                    isSigner: true,
                },
                {
                    name: "systemProgram",
                    isMut: false,
                    isSigner: false,
                },
            ],
            args: [
                {
                    name: "extcallId",
                    type: {
                        array: ["u8", 32],
                    },
                },
            ],
        },
        {
            name: "executeExtcall",
            accounts: [
                {
                    name: "extcallMeta",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "extcallStorage",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "executeWallet",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "executeAuth",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "executor",
                    isMut: true,
                    isSigner: true,
                },
                {
                    name: "rewardDst",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "splTokenProgram",
                    isMut: false,
                    isSigner: false,
                },
                {
                    name: "systemProgram",
                    isMut: false,
                    isSigner: false,
                },
            ],
            args: [
                {
                    name: "extcallId",
                    type: {
                        array: ["u8", 32],
                    },
                },
                {
                    name: "count",
                    type: "u64",
                },
                {
                    name: "reversedSubsitutionBumps",
                    type: "bytes",
                },
            ],
        },
        {
            name: "makeFallback",
            accounts: [
                {
                    name: "extcallMeta",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "executeAuth",
                    isMut: false,
                    isSigner: false,
                },
                {
                    name: "executeWallet",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "fallbackAuthority",
                    isMut: true,
                    isSigner: true,
                },
                {
                    name: "fallbackWallet",
                    isMut: true,
                    isSigner: false,
                },
                {
                    name: "systemProgram",
                    isMut: false,
                    isSigner: false,
                },
                {
                    name: "splTokenProgram",
                    isMut: false,
                    isSigner: false,
                },
            ],
            args: [
                {
                    name: "extcallId",
                    type: {
                        array: ["u8", 32],
                    },
                },
            ],
        },
    ],
    accounts: [
        {
            name: "extcallMeta",
            type: {
                kind: "struct",
                fields: [
                    {
                        name: "extcallId",
                        type: {
                            array: ["u8", 32],
                        },
                    },
                    {
                        name: "state",
                        type: {
                            defined: "ExtcallMetaState",
                        },
                    },
                    {
                        name: "hash",
                        type: {
                            array: ["u8", 32],
                        },
                    },
                    {
                        name: "providedAmount",
                        type: "u64",
                    },
                    {
                        name: "providedToken",
                        type: {
                            defined: "ProvidedToken",
                        },
                    },
                    {
                        name: "providedAuthority",
                        type: "publicKey",
                    },
                    {
                        name: "bump",
                        type: "u8",
                    },
                    {
                        name: "executeWalletBump",
                        type: "u8",
                    },
                    {
                        name: "executeAuthBump",
                        type: "u8",
                    },
                ],
            },
        },
        {
            name: "state",
            type: {
                kind: "struct",
                fields: [
                    {
                        name: "protocolAuthority",
                        type: "publicKey",
                    },
                    {
                        name: "whitelist",
                        type: {
                            vec: "publicKey",
                        },
                    },
                    {
                        name: "bump",
                        type: "u8",
                    },
                ],
            },
        },
        {
            name: "extcallStorage",
            type: {
                kind: "struct",
                fields: [
                    {
                        name: "instructionsCount",
                        type: "u64",
                    },
                    {
                        name: "blocked",
                        type: "bool",
                    },
                    {
                        name: "externalCallData",
                        type: "bytes",
                    },
                ],
            },
        },
    ],
    types: [
        {
            name: "ProvidedToken",
            type: {
                kind: "enum",
                variants: [
                    {
                        name: "NativeSol",
                    },
                    {
                        name: "SplToken",
                    },
                ],
            },
        },
        {
            name: "ExtcallMetaState",
            type: {
                kind: "enum",
                variants: [
                    {
                        name: "Initializing",
                    },
                    {
                        name: "Execution",
                        fields: [
                            {
                                name: "offset",
                                type: "u64",
                            },
                            {
                                name: "buffer",
                                type: "publicKey",
                            },
                            {
                                name: "executed_instructions_count",
                                type: "u64",
                            },
                        ],
                    },
                    {
                        name: "Executed",
                        fields: [
                            {
                                name: "buffer",
                                type: "publicKey",
                            },
                        ],
                    },
                    {
                        name: "Failed",
                        fields: [
                            {
                                name: "buffer",
                                type: "publicKey",
                            },
                        ],
                    },
                    {
                        name: "Fallbacked",
                    },
                ],
            },
        },
    ],
    events: [
        {
            name: "Executed",
            fields: [
                {
                    name: "isFully",
                    type: "bool",
                    index: false,
                },
                {
                    name: "presentlyExecutedInstructionsCount",
                    type: "u64",
                    index: false,
                },
                {
                    name: "executedInstructionsCount",
                    type: "u64",
                    index: false,
                },
                {
                    name: "totalInstructionsCount",
                    type: "u64",
                    index: false,
                },
            ],
        },
        {
            name: "Fallbacked",
            fields: [
                {
                    name: "executedInstructionsCount",
                    type: "u64",
                    index: false,
                },
            ],
        },
    ],
    errors: [
        {
            code: 6000,
            name: "NotAllowedCreator",
        },
        {
            code: 6001,
            name: "NativeSolLiquiditySourcePubkeysNotMatch",
            msg: "For NativeSol flow the pubkeys `LiquiditySourceAuth` & `LiquiditySourceWallet` must match",
        },
        {
            code: 6002,
            name: "NativeSolWrongExecutorWallet",
            msg: "For NativeSol flow pubkey of ExecutorWallet must be equal to ExtcallAuth",
        },
        {
            code: 6003,
            name: "SplWrongTokenMint",
            msg: "For SplFlow token mint should be consistence",
        },
        {
            code: 6004,
            name: "SplWrongExtcallWallet",
            msg: "For SplFlow wrong extcall wallet",
        },
        {
            code: 6005,
            name: "WrongExtcallWallet",
            msg: "Wrong extcall wallet, please check flow rules",
        },
        {
            code: 6006,
            name: "NotExpectedExtcallStateForExecution",
            msg: "For execute, state of extcall meta must be `Initializing` or `Execution`",
        },
        {
            code: 6007,
            name: "NotExpectedExtcallStateForFallback",
            msg: "For fallback, state of extcall meta must be `Initializing` or `Execution`",
        },
        {
            code: 6008,
            name: "WrongHashOfExtcall",
        },
        {
            code: 6009,
            name: "WrongExtcallStorage",
        },
        {
            code: 6010,
            name: "ExecutionFinishedInMandatoryBlock",
        },
        {
            code: 6011,
            name: "CantCloseValidatedExtcallStorage",
        },
        {
            code: 6012,
            name: "CorruptedExtcallStorage",
        },
        {
            code: 6013,
            name: "PreparingAccountError",
        },
        {
            code: 6014,
            name: "ExtcallStorageBlocked",
        },
    ],
};
