// @debridge-finance/dln-client/dist/esm/evm/core/services/crosschain-forwarder.service.js
import { encodeFunctionData } from "viem";
import { CrosschainForwarderABI } from "../../assets/abi/CrossChainForwarder";
export class CrossChainForwarderService {
    strictlySwapAndCallDln(payload) {
        return encodeFunctionData({
            abi: CrosschainForwarderABI,
            functionName: "strictlySwapAndCallDln",
            args: [
                //_srcTokenIn
                payload._srcTokenIn,
                //_srcAmountIn
                payload._srcAmountIn,
                //_srcTokenInPermitEnvelope
                payload._srcTokenInPermitEnvelope,
                //_swapDetails
                {
                    swapRouter: payload._srcSwapRouter,
                    swapCalldata: payload._srcSwapCalldata,
                    tokenOut: payload._srcTokenOut,
                    tokenOutExpectedAmount: payload._srcTokenExpectedAmountOut,
                    // refund from fulfill's preswap is sent to the user (the maker) to limit taker's income
                    tokenOutRefundRecipient: payload._srcTokenRefundRecipient,
                },
                //_target
                payload._receiver,
                //_targetData
                payload._receiverCalldata,
                //_orderId
                payload._orderId,
            ],
        });
    }
    strictlySwapAndCall(provider, contractAddress, payload) {
        return encodeFunctionData({
            abi: CrosschainForwarderABI,
            functionName: "strictlySwapAndCall",
            args: [
                //_srcTokenIn
                payload._srcTokenIn,
                //_srcAmountIn
                payload._srcAmountIn,
                //_srcTokenInPermitEnvelope
                payload._srcTokenInPermitEnvelope,
                //_swapRouter
                payload._srcSwapRouter,
                //_swapCalldata
                payload._srcSwapCalldata,
                //_tokenOut
                payload._srcTokenOut,
                //_tokenOutExpectedAmount
                payload._srcTokenExpectedAmountOut,
                // refund from fulfill's preswap is sent to the user (the maker) to limit taker's income
                //_tokenOutRefundRecipient
                payload._srcTokenRefundRecipient,
                //_target
                payload._receiver,
                //_targetData
                payload._receiverCalldata,
            ],
        });
    }
}
