// @debridge-finance/dln-client/dist/esm/swapConnector/neon-usdc-lp.connector.js
import { ChainId } from "../common.types";
import { tokenStringToBuffer } from "../utils";
import { UniswapV2Connector } from "./uniswapv2.connector";
export const NeonTokenAddresses = {
    UsdcAddress: "0xEA6B04272f9f62F997F666F07D3a974134f7FFb9",
    UsdtAddress: "0x5f0155d08ef4aae2b500aefb64a3419da8bb611a",
    wSolAddress: "0x5f38248f339Bf4e84A2caf4e4c0552862dC9F82a",
    wNeonAddress: "0x202C35e517Fa803B537565c40F0a6965D7204609",
    MoraSwapV2RouterAddress: "0x594e37B9F39f5D31DEc4a8c1cC4fe2E254153034",
    ICSV2RouterAddress: "0xBb5e1777A331ED93E07cF043363e48d320eb96c4",
};
export const NeonTokenAddressBuffers = {
    UsdcAddress: tokenStringToBuffer(ChainId.Neon, NeonTokenAddresses.UsdcAddress),
    UsdtAddress: tokenStringToBuffer(ChainId.Neon, NeonTokenAddresses.UsdtAddress),
    wSolAddress: tokenStringToBuffer(ChainId.Neon, NeonTokenAddresses.wSolAddress),
    wNeonAddress: tokenStringToBuffer(ChainId.Neon, NeonTokenAddresses.wNeonAddress),
    NeonAddress: Buffer.alloc(20, 0),
    MoraSwapV2RouterAddress: tokenStringToBuffer(ChainId.Neon, NeonTokenAddresses.MoraSwapV2RouterAddress),
    ICSV2RouterAddress: tokenStringToBuffer(ChainId.Neon, NeonTokenAddresses.ICSV2RouterAddress),
};
export class NeonUsdcLp extends UniswapV2Connector {
    constructor(connection) {
        super(connection, ChainId.Neon, [
            [NeonTokenAddressBuffers.UsdcAddress, NeonTokenAddressBuffers.wSolAddress],
            [NeonTokenAddressBuffers.wSolAddress, NeonTokenAddressBuffers.wNeonAddress],
        ], NeonTokenAddressBuffers.MoraSwapV2RouterAddress, NeonTokenAddressBuffers.wNeonAddress);
    }
}
export class NeonMoraswap extends NeonUsdcLp {
}
export class NeonICS extends UniswapV2Connector {
    constructor(connection) {
        super(connection, ChainId.Neon, [
            // USDC/USDT
            [NeonTokenAddressBuffers.UsdcAddress, NeonTokenAddressBuffers.UsdtAddress],
            // USDT/wSOL
            [NeonTokenAddressBuffers.wSolAddress, NeonTokenAddressBuffers.UsdtAddress],
            // USDT/wNEON
            [NeonTokenAddressBuffers.wNeonAddress, NeonTokenAddressBuffers.UsdtAddress],
        ], NeonTokenAddressBuffers.ICSV2RouterAddress, NeonTokenAddressBuffers.wNeonAddress);
    }
}
