// @debridge-finance/dln-client/dist/esm/swapConnector/swap.connector.js
var _SwapConnector_instances, _SwapConnector_fetch;
import { __classPrivateFieldGet } from "tslib";
import { BPS_DENOMINATOR, } from "../common.types";
import { SwapConnectorError } from "./errors/swap-connector.error";
import { SwapConnectorErrorType } from "./errors/swap-connector-error.type";
import { ErrorResponse } from "./errors/error-response";
export function bpsToPercent(bps) {
    return (bps * 100) / Number(BPS_DENOMINATOR);
}
export class SwapConnector {
    constructor() {
        _SwapConnector_instances.add(this);
    }
    async call(params, context) {
        const basicUrl = params.basicUrl.endsWith("/")
            ? params.basicUrl.slice(0, params.basicUrl.length - 1)
            : params.basicUrl;
        const url = `${basicUrl}${params.path}?${params.query.toString()}`;
        const response = await __classPrivateFieldGet(this, _SwapConnector_instances, "m", _SwapConnector_fetch).call(this, url, {
            headers: params.headers,
            body: params.body,
            method: params.method,
            signal: params.timeoutInMs ? AbortSignal.timeout(params.timeoutInMs) : undefined,
        }, context.logger);
        const data = await response.text();
        if (!response.ok) {
            throw new ErrorResponse(response.status, response.statusText || `Request is failed`, data);
        }
        return JSON.parse(data);
    }
}
_SwapConnector_instances = new WeakSet(), _SwapConnector_fetch = async function _SwapConnector_fetch(url, request, logger) {
    const parsedUrl = new URL(url);
    const reqCtx = {
        url: parsedUrl.href,
        pathname: parsedUrl.pathname,
        host: parsedUrl.host,
        request,
        type: "external_request",
    };
    logger.log(reqCtx, `Swap request`);
    const startedAt = new Date().getTime();
    let response;
    try {
        response = await fetch(url, request);
        const resp = (await response.clone().json()) ?? null;
        const resCtx = {
            ...reqCtx,
            response: resp,
            duration: Date.now() - startedAt,
        };
        logger.log(resCtx, `Swap response`);
    }
    catch (e) {
        const error = e;
        const errorCtx = {
            ...reqCtx,
            error,
            duration: Date.now() - startedAt,
        };
        logger.error(errorCtx, `Swap request was failed`);
        const errorMessage = `${this.connectorName}::request failed: ${e}`;
        throw new SwapConnectorError(this.connectorName, SwapConnectorErrorType.InternalError, errorMessage);
    }
    return response;
};
