// @debridge-finance/dln-client/dist/esm/errors/client-error-type.enum.js
export var ClientErrorType;
(function (ClientErrorType) {
    ClientErrorType["UnsupportedChain"] = "UnsupportedChain";
    ClientErrorType["WrongSolanaAddressLength"] = "WrongSolanaAddressLength";
    ClientErrorType["ErrorEqualChainsForCrossChainSwaps"] = "ErrorEqualChainsForCrossChainSwaps";
    ClientErrorType["ErrorLowGiveAmount"] = "ErrorLowGiveAmount";
    ClientErrorType["SolanaNoSrcState"] = "SolanaNoSrcState";
    ClientErrorType["SolanaWrongSrcChainId"] = "WrongSrcChainId";
    ClientErrorType["WrongOrderId"] = "WrongOrderId";
    ClientErrorType["UnderestimatedSwap"] = "UnderestimatedSwap";
    ClientErrorType["Connector1inchReturnedError"] = "Connector1inchReturnedError";
    ClientErrorType["Connector1inchFailed"] = "Connector1inchFailed";
    ClientErrorType["FakeOrderFulfillEstimationFailed"] = "FakeOrderFulfillEstimationFailed";
    ClientErrorType["ZeroEstimatorBalance"] = "ZeroEstimatorBalance";
    ClientErrorType["JupiterError"] = "JupiterError";
    ClientErrorType["NoSwapRoutes"] = "NoSwapRoutes";
    ClientErrorType["UnsupportedFinalization"] = "UnsupportedFinalization";
    ClientErrorType["GiveStateNotFound"] = "GiveStateNotFound";
    ClientErrorType["BadALT"] = "BadALT";
    ClientErrorType["WrongChain"] = "WrongChain";
    ClientErrorType["SwapConnectorError"] = "SwapConnectorError";
    ClientErrorType["WrongBatchSize"] = "WrongBatchSize";
    ClientErrorType["UnknownStatus"] = "UnknownStatus";
    ClientErrorType["UnkownEngine"] = "UnkownEngine";
    ClientErrorType["ClientMissing"] = "ClientMissing";
    ClientErrorType["ConnectionMissing"] = "ConnectionMissing";
    ClientErrorType["PropertyMissing"] = "PropertyMissing";
    ClientErrorType["FailedTransaction"] = "FailedTransaction";
    ClientErrorType["MissingTransaction"] = "MissingTransaction";
    ClientErrorType["SolanaWrongOrderState"] = "SolanaWrongOrderState";
    ClientErrorType["RewardsRequired"] = "RewardsRequired";
    ClientErrorType["WrongBatchOrders"] = "WrongBatchOrders";
    ClientErrorType["RewardBeneficiaryRequired"] = "RewardBeneficiaryRequired";
    ClientErrorType["AuthroizedNativeSenderAccountMissing"] = "AuthroizedNativeSenderAccountMissing";
})(ClientErrorType || (ClientErrorType = {}));
